components/server/src/api/envvar-service-api.ts

```diff
@@ -0,0 +1,243 @@
+
+    @inject(WorkspaceService)
+    private readonly workspaceService: WorkspaceService;
+
+    @inject(PublicAPIConverter)
+    private readonly apiConverter: PublicAPIConverter;
+
+    async listUserEnvironmentVariables(
+        req: ListUserEnvironmentVariablesRequest,
+        context: HandlerContext,
+    ): Promise<ListUserEnvironmentVariablesResponse> {
+        const response = new ListUserEnvironmentVariablesResponse();
+        const userEnvVars = await this.envVarService.listUserEnvVars(context.user.id, context.user.id);
+        response.environmentVariables = userEnvVars.map((i) => this.apiConverter.toUserEnvironmentVariable(i));
+
+        return response;
+    }
+
+    async updateUserEnvironmentVariable(
+        req: UpdateUserEnvironmentVariableRequest,
+        context: HandlerContext,
+    ): Promise<UpdateUserEnvironmentVariableResponse> {
+        if (!req.envVarId) {
+            throw new ConnectError("envVarId is not set", Code.InvalidArgument);
+        }
+
+        const response = new UpdateUserEnvironmentVariableResponse();
+
+        const userEnvVars = await this.envVarService.listUserEnvVars(context.user.id, context.user.id);
+        const userEnvVarfound = userEnvVars.find((i) => i.id === req.envVarId);
+        if (userEnvVarfound) {
+            const variable: UserEnvVarValue = {
+                name: req.name ?? userEnvVarfound.name,
+                value: req.value ?? userEnvVarfound.value,
+                repositoryPattern: req.repositoryPattern ?? userEnvVarfound.repositoryPattern,
+            };
+            variable.repositoryPattern = UserEnvVar.normalizeRepoPattern(variable.repositoryPattern);
+
+            await this.envVarService.updateUserEnvVar(context.user.id, context.user.id, variable);
+
+            const updatedUserEnvVars = await this.envVarService.listUserEnvVars(context.user.id, context.user.id);
+            const updatedUserEnvVar = updatedUserEnvVars.find((i) => i.id === req.envVarId);
+            if (!updatedUserEnvVar) {
+                throw new ConnectError("could not update env variable", Code.Internal);
+            }
+
+            response.environmentVariable = this.apiConverter.toUserEnvironmentVariable(updatedUserEnvVar);
+            return response;
+        }
+
+        throw new ConnectError("env variable not found", Code.InvalidArgument);
+    }
+
+    async createUserEnvironmentVariable(
+        req: CreateUserEnvironmentVariableRequest,
+        context: HandlerContext,
+    ): Promise<CreateUserEnvironmentVariableResponse> {
+        const response = new CreateUserEnvironmentVariableResponse();
+
+        const variable: UserEnvVarValue = {
+            name: req.name,
+            value: req.value,
+            repositoryPattern: req.repositoryPattern,
+        };
+        variable.repositoryPattern = UserEnvVar.normalizeRepoPattern(variable.repositoryPattern);
+
+        await this.envVarService.addUserEnvVar(context.user.id, context.user.id, variable);
```

Comment:
As far as I understand typeorm has repo.save method which will create, update and return new object.